crypto/getenv.c

```diff
@@ -13,9 +13,45 @@
 
 #include <stdlib.h>
 #include "internal/cryptlib.h"
+#include "e_os.h"
 
 char *ossl_safe_getenv(const char *name)
 {
+#if (defined(OPENSSL_SYS_WINDOWS)) && defined(CP_UTF8) && !defined(_WIN32_WCE)
+
+    if (GetEnvironmentVariableW(L"OPENSSL_WIN32_UTF8", NULL, 0) == 0)
+        return getenv(name);
+
+    /* convert name to wide string */
+    int rsize = mbstowcs(NULL, name, 0) + 1;
+    WCHAR * namew = _alloca(rsize * sizeof(WCHAR));
+    int fsize = mbstowcs(namew, name, rsize);
+
+    char *val = NULL;
+
+    /* determine value string size in wchars */
+    DWORD envlen = GetEnvironmentVariableW(namew, NULL, 0);    
```

Comment:
this does not include the NULL character so your _alloc is off by 1.

Reply:
MS DOC:

If lpBuffer is not large enough to hold the data, the return value is the buffer size, in characters, required to hold the string and its terminating null character and the contents of lpBuffer are undefined.

Reply:
Strange.. i was also reading microsoft docs
https://docs.microsoft.com/en-us/windows/win32/api/processenv/nf-processenv-getenvironmentvariablew

Reply:
working example:

env TEST=12345

GetEnvironmentVariableW('TEST', NULL, 0) return 6

6 = 5 characters + NULL terminator

so allocated buffer has space for env value and NULL terminator

where is problem?

Reply:
@slontis, from that page (bold mine):

> **If the function succeeds**, the return value is the number of characters stored in the buffer pointed to by _lpBuffer_, **not including the terminating null character**.
> 
> **If _lpBuffer_ is not large enough to hold the data**, the return value is the buffer size, in characters, **required to hold the string and its terminating null character** and the contents of _lpBuffer_ are undefined.

Leave it to Microsoft to give mixed messages....